File: day4/task1.py
def check(matr: list[str], x: int, y: int, i: int, j: int, word: str) -> bool:
    curr = 0
    while curr < len(word) and 0 <= x < len(matr) and 0 <= y < len(matr[0]):
        if word[curr] != matr[x][y]:
            break
        x += i
        y += j
        curr += 1
    return curr == len(word)


with open("input.txt") as f:
    lines = f.readlines()

res = sum(
    check(lines, x, y, i, j, "XMAS")
    for x in range(len(lines))
    for y in range(len(lines[0]))
    for i in range(-1, 2)
    for j in range(-1, 2)
    if i or j
)
print(res)

File: day4/task1.cpp
#include <bits/stdc++.h>
using namespace std;

bool check(const vector<string> &matr, int x, int y, int i, int j,
           string word) {
    size_t curr{0};
    while (curr < word.length() && x == clamp(x, 0, (int)matr.size() - 1) &&
           y == clamp(y, 0, (int)matr[0].length())) {
        if (word[curr] != matr[x][y])
            break;
        x += i;
        y += j;
        ++curr;
    }
    return word.length() == curr;
}

int main(void) {
    vector<string> lines;
    string line;
    ifstream file{"input.txt"};
    while (file >> line)
        lines.push_back(line);
    file.close();

    namespace views = ranges::views;
    auto rng_x = views::iota(0, (int)lines.size());
    auto rng_y = views::iota(0, (int)lines[0].size());
    auto rng_i = views::iota(-1, 2);
    auto rng_j = views::iota(-1, 2);
    auto cartesian = views::cartesian_product(rng_x, rng_y, rng_i, rng_j);
    auto filtered = cartesian | views::filter([](const auto &tpl) {
                        auto [_, _, i, j] = tpl;
                        return i || j;
                    });
    auto transformed = views::transform(filtered, [&lines](const auto &tpl) {
        auto [x, y, i, j] = tpl;
        return check(lines, x, y, i, j, "XMAS");
    });
    int res{accumulate(transformed.begin(), transformed.end(), 0)};

    cout << res << '\n';
    return 0;
}
